File: Editorial/0117_Populating_Next_Right_Pointers_in_Each_Node_II/Go/1_Level_Order_Traversal.go
func connect(root *Node) *Node {
    if root == nil {
        return root
    }
    // Initialize a queue data structure which contains
    // just the root of the tree
    Q := list.New()
    Q.PushBack(root)
    // Outer while loop which iterates over
    // each level
    for Q.Len() > 0 {
        // Note the size of the queue
        size := Q.Len()
        // Iterate over all the nodes on the current level
        for i := 0; i < size; i++ {
            // Pop a node from the front of the queue
            front := Q.Front()
            node := front.Value.(*Node)
            Q.Remove(front)
            // This check is important. We don't want to
            // establish any wrong connections. The queue will
            // contain nodes from 2 levels at most at any
            // point in time. This check ensures we only
            // don't establish next pointers beyond the end
            // of a level
            if i < size-1 {
                node.Next = Q.Front().Value.(*Node)
            }
            // Add the children, if any, to the back of
            // the queue
            if node.Left != nil {
                Q.PushBack(node.Left)
            }
            if node.Right != nil {
                Q.PushBack(node.Right)
            }
        }
    }
    // Since the tree has now been modified, return the root node
    return root
}

File: Editorial/0101-0200/0117_Populating_Next_Right_Pointers_in_Each_Node_II/TypeScript/1_Level_Order_Traversal.ts
function connect(root: Node | null): Node | null {
    if (!root) return root;
    // Initialize a Queue with the root
    const Q: Node[] = [];
    Q.push(root);
    // Loop while the Queue is not empty
    while (Q.length > 0) {
        const size = Q.length;
        // Iterate over all the nodes currently in the Queue
        for (let i = 0; i < size; i++) {
            // Remove the first node from the Queue
            const node: Node | undefined = Q.shift();
            // If this is not the last node in the level, point next to the next node
            if (i < size - 1) {
                if (node) node.next = Q[0];
            }
            // Add the left and right children (if they exist) to the Queue
            if (node?.left) Q.push(node?.left);
            if (node?.right) Q.push(node?.right);
        }
    }
    // Return the root
    return root;
}
